From a17fb47e08257d09e1e795e8114b21b134b1f3c2 Mon Sep 17 00:00:00 2001
From: Owen W. Taylor <otaylor@fishsoup.net>
Date: Mon, 19 Oct 2009 18:43:00 -0400
Subject: [PATCH] Don't focus ancestor window on a different workspace

When we are moving a window with a modal dialog to a different
workspace, meta_workspace_focus_default_window() can be called
with 'not_this_one' being the focused modal dialog.

Since the ancestor of that window is also being moved, we must
not focus it as an alternative to the current window; this will
cause windows to be moved back and Metacity to get into an
inconsistent confused state.

https://bugzilla.redhat.com/show_bug.cgi?id=237158

https://bugzilla.gnome.org/show_bug.cgi?id=598995
---
diff -ur metacity-2.30.3-ylmf1/src/core/workspace.c metacity-2.30.3-ylmf1.new//src/core/workspace.c
--- metacity-2.30.3-ylmf1/src/core/workspace.c	2010-09-05 00:09:53.000000000 +0800
+++ metacity-2.30.3-ylmf1.new//src/core/workspace.c	2011-09-29 17:41:06.533276306 +0800
@@ -970,8 +970,11 @@
       MetaWindow *ancestor;
       ancestor = NULL;
       meta_window_foreach_ancestor (not_this_one, record_ancestor, &ancestor);
-      if (ancestor != NULL)
-        {
+      if (ancestor != NULL &&
+           (ancestor->on_all_workspaces ||
+            ancestor->workspace == workspace) &&
+            meta_window_showing_on_its_workspace (ancestor))
+	{
           meta_topic (META_DEBUG_FOCUS,
                       "Focusing %s, ancestor of %s\n", 
                       ancestor->desc, not_this_one->desc);
